### DISCUSSION TITLE
IoC Container / Bean 기본 개념

### DISCUSSION THREAD

**Discussion by seunghye218** [2024-woowacourse-study/level-interview#148] [Spring Core]
.

**Comment by Kimprodp** [comment]
**[IoC Container]**
스프링 프레임워크에는 사용할 객체들을 미리 생성하고 의존성을 관리해주는 컨테이너가 있다.
이 컨테이너를 통해 생성된 객체를 사용할 곳에 주입 시켜 사용 가능하게 함으로써 객체의 의존성을 역전 시켜 객체 간의 결합도를 줄이고, 유연한 코드를 작성할 수 있게 한다. 이 컨테이너를 '스프링 컨테이너' 또는 'IoC 컨테이너' 라고 한다.

**[Bean]**
Bean은 스프링 컨테이너에서 인스턴스화 된 객체를 의미하며, 스프링 컨테이너에 의해 관리되는 재사용 가능한 객체이다. 즉, 스프링 컨테이너에 등록된 객체를 스프링 빈이라고 할 수 있으며 스프링 컨테이너가 관리하는 자바 객체라는 뜻이다.

**Comment by seunghye218** [comment]
Spring Framework의 핵심 요소로, 애플리케이션의 객체 생명 주기와 의존성 관리에 중요한 역할을 합니다. 객체지향 설계의 원칙을 잘 지원하여 유지보수성과 확장성을 높이는 데 도움을 줍니다.

## IoC 컨테이너 (Inversion of Control Container)

IoC 컨테이너는 Spring Framework의 핵심 구성 요소로, 애플리케이션에서 객체의 생성, 구성, 생명 주기를 관리하는 역할을 합니다. IoC 컨테이너는 의존성 주입(Dependency Injection)을 통해 객체 간의 의존성을 설정하고 관리합니다.

- 주요 기능
1. 객체 생성 및 관리: 애플리케이션에서 필요한 객체를 생성하고 관리합니다.
2. 의존성 주입: 객체 간의 의존성을 설정하고 주입합니다.
3. 생명 주기 관리: 빈(Bean)의 생명 주기를 관리합니다.

- 제어의 역전 (Inversion of Control): IoC의 주요 원칙으로 객체의 생성과 의존성 관리의 책임이 애플리케이션 코드가 아닌 컨테이너로 넘어갑니다. 이를 통해 코드가 더 유연하고 테스트 가능하게 됩니다.

## Bean

Bean은 Spring IoC 컨테이너에 의해 관리되는 객체를 의미합니다. 빈은 컨테이너가 생성하고 초기화하며, 필요에 따라 의존성을 주입받습니다.

- 주요 특징
1. 컨테이너 관리: 빈은 IoC 컨테이너에 의해 생성되고 관리됩니다.
2. 의존성 주입: 빈 간의 의존성은 컨테이너가 주입합니다.
3. 생명 주기: 빈의 생명 주기는 컨테이너가 관리하며, 필요한 경우 초기화 및 소멸 콜백 메서드를 호출할 수 있습니다.

- Bean의 생명 주기
1. 객체 생성: 빈이 컨테이너에 의해 생성됩니다.
2. 의존성 주입: 생성된 빈에 필요한 의존성이 주입됩니다.
3. 초기화: 빈이 초기화되며, 필요한 경우 초기화 메서드가 호출됩니다.
4. 사용: 빈이 애플리케이션 로직에서 사용됩니다.
5. 소멸: 애플리케이션 종료 시 빈이 소멸되며, 필요한 경우 소멸 메서드가 호출됩니다.
